crates/gitbutler-git/src/executor/tokio/windows.rs

```diff
@@ -10,30 +10,25 @@ use tokio::{
     net::windows::named_pipe::{NamedPipeServer, ServerOptions},
     sync::Mutex,
 };
+use windows::Win32::{Foundation::HANDLE, System::Pipes::GetNamedPipeClientProcessId};
 
-const ASKPASS_PIPE_PREFIX: &str = r"\\.\pipe\gitbutler-askpass-";
+// Slashes instead of backslashes to prevent any issues with escaping.
+const ASKPASS_PIPE_PREFIX: &str = r"//./pipe/gitbutler-askpass-";
```

Comment:
Out of curiosity, does this have to be a named pipe? Can you capture askpass's stdout instead? Any time a pipe has an explicit name you now have to do a lot of Security Thinking about who can access this pipe, and what happens if someone Bad can mess with it

Reply:
The `nMaxInstances` parameter when `CreateNamedPipeW` is called could limit it to only one connection, but this is probably not sufficient for security, and I will need to check the Tokio internals for it.

There is also a permission mechanism in Windows to only allow a specific process handle to connect.
I don't know the details yet, but it's possible.

However, the executable of this process (gitbutler-git-askpass.exe) could also be replaced. 
So, we will need to check the signature of this executable before calling it.

I don't quite understand how to capture the standard output. I am not aware of a solution that could capture only part of the standard output, as Git will interpret any output as a password.  
Could you please elaborate on this? It could be a great solution.

As for non-named pipes, they will be more secure, but I am not sure if GitButler will still be considered the parent process since it will be spawned by Git.

I am sorry I didn't think this through initially. I was only trying to make the current implementation work. :/

Reply:
> I don't quite understand how to capture the standard output. I am not aware of a solution that could capture only part of the standard output, as Git will interpret any output as a password.

Yeah, I don't quite know how GitButler is interacting with askpass, so this might not be workable, I was more writing generally - https://stackoverflow.com/questions/32641423/how-to-access-an-inherited-anonymous-pipe-handle-other-than-stdout-stderr-st has some useful details around anonymous pipes, though if Git is spawning -askpass it might not work, you can always play around with it. 

> I am sorry I didn't think this through initially. I was only trying to make the current implementation work. :/

No worries at all! This can always be improved in a later PR

Reply:
Could you briefly explain how you secured the Linux side, if you don't mind? 
If this could apply to Windows, it would be a great start for me ! 

I will try to mess around with software signatures and non-named pipes, if possible, or fall back to named pipes with extra security.
Is it needed to check the git signature from various windows providers or is it out of scope and prone to errors ? 

Kinda lost if you could limit my scope this will be great !

There is plenty of stuff i dont know so i cant guarantee my success but i will definitely try in the upcoming days or maybe weeks !

Would love to contribute to the project

Reply:
I did not refresh the page so i just saw your answer. I will have a look thanks

Reply:
> Yeah, I don't quite know how GitButler is interacting with askpass, so this might not be workable, I was more writing generally - https://stackoverflow.com/questions/32641423/how-to-access-an-inherited-anonymous-pipe-handle-other-than-stdout-stderr-st has some useful details around anonymous pipes, though if Git is spawning -askpass it might not work, you can always play around with it.

I can't pass any arguments to gitbutler-git-askpass or even write to its standard input. However, I could take inspiration from it and pass the non-named-pipe handle as an environment variable, but I don't know the scope of environment variables when initialized inside Gitbutler for security purposes.
I'm also unsure if the handle inheritance will work (as it's started by Git, not Gitbutler).

Should I take this approach?

Or should I keep using the named pipe? However, I don't see how I could get the right process handle to limit the permission of the pipe to it (since it's started by Git, not Gitbutler).

Alternatively, should I just limit the number of connections to the named pipe to 1 ?
As gitbutler-git-askpass will probably be the first to connect to it as the pipe name at the end is randomly generated.

How did you handle this security concern of the linux side ? Maybe i could try to reproduce it.

## Facultative reading
I think there is too many way a bad actor could mess with any of the following systems.
Process injection, Manual mapping, replacing windows api functions, git, gitbutler-askpass or even a simple Keylogger and many more ... 

I also want this process to be secure as possible but even simple attack vectors will just work on any solution mentioned :/

Unfortunatly i feel like on windows there not much we can do as a simple keylogger using the standard library doesn't even require admin privilege. (This is not the case on linux)

The best we can do is assume the AV will do a proper work and trust windows functions.

But i agree the current solution need at least an extrat step. (As i dont think we limit the pipe acces strongly egough)

Sorry for this reply i just look like a depressed dev 🤠

Reply:
Hey !

I started implementing the anonymous pipe and i feels kind wrong to convert a pointer to string and vice versa.
This introduce some usage overhead (make sure the pointer is not a dangling pointer due to the rust borrowing rules ect). Even if for now it work (I havent checked the process inheritance when spawned by git).

I feel like it doesn't add more than limiting the pipe connection to only 1 instance.
As if a bad actor connect to the pipe first i will not have the secret passed in the scoped environment and will not receive the password as such.

Let me know if this it's still your preferred solution and i will keep working on it.

Reply:
So i tried to use a non named pipe and i keep getting access violation errors as soon as the process is created with rust. The bypass would be to create the process directly with CreateProcess and the bInheritHandles option on windows.
But we also got no warranty that git will create the process with same option.

So i guess i will just see how to limit the number of named pipe connection with tokio.